### PR TITLE
Sub-issue (2721): ADSR/GAIN envelopes, KON/KOFF, noise, and PMON

### PR DESCRIPTION
**Pull request by rmstdope**
## Summary
- implement #2777 S-DSP behavior for ADSR/GAIN envelopes, KON/KOFF handling, NON noise routing, PMON pitch modulation, and ENVX/OUTX updates
- refactor SNES DSP implementation from `src/snes/apu/dsp.rs` into focused submodules under `src/snes/apu/dsp/`
- add dedicated DSP tests for envelope progression, KON/KOFF latency edge cases, noise sequencing, PMON behavior, and legacy-compatible deserialize/normalize flow
- update `architecture.md` to reflect the new SNES DSP module layout and behavior

## Details
- Added submodules:
  - `src/snes/apu/dsp/mod.rs`
  - `src/snes/apu/dsp/voice.rs`
  - `src/snes/apu/dsp/envelope.rs`
  - `src/snes/apu/dsp/brr.rs`
  - `src/snes/apu/dsp/gaussian.rs`
  - `src/snes/apu/dsp/tests.rs`
- Added per-voice state model with serde-defaulted additive fields and restore normalization to keep save-state compatibility.
- Implemented KON/KOFF interaction fix where KOFF cancels pending KON delay.
- Implemented PMON using pre-mixer modulation source (independent from master volume).

## Validation
- `cargo clippy --all-targets --all-features -- -D warnings`
- `cargo fmt`
- `./scripts/test-dir.sh src/snes/apu --skip-integration`
- `cargo test --no-default-features --lib`
- `wasm-pack test --headless --chrome --no-default-features --features wasm`
- `source .venv/bin/activate && python -m unittest discover -s scripts -t scripts -p "test_*.py"`
- `source .venv/bin/activate && python -m unittest discover -s scripts/metadata-scraper -t scripts/metadata-scraper -p "test_*.py"`
- `source .venv/bin/activate && python -m unittest discover -s scripts/nes-rom-db-scraper -t scripts/nes-rom-db-scraper -p "test_*.py"`
- `npm test`

Closes #2777

### CODE REVIEW

File: src/snes/apu/dsp/mod.rs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+            self.voices[voice].sample_pos = self.voices[voice]
+                .sample_pos
+                .wrapping_add(u32::from(effective_pitch));
+            let sample = self.voice_sample(voice, non);
+            self.voices[voice].mod_source = (sample >> 8).clamp(-128, 127) as i8;
+            let (left, right) = self.mix_voice_sample(voice, sample);
+            let mixed = (((i32::from(left) + i32::from(right)) / 2) >> 8).clamp(-128, 127) as i8;
+            self.voices[voice].outx = mixed;
+            self.regs[(voice << 4) + 8] = self.voices[voice].envx;
+            self.regs[(voice << 4) + 9] = mixed as u8;
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/mod.rs]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/mod.rs [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        envelope::step_voice_envelope(v);
+    }
+
+    pub fn write_reg(&mut self, addr: u8, value: u8) {
+        let reg = addr & 0x7F;
+        let index = usize::from(reg);
+        if index >= self.regs.len() {
+            return;
+        }
+        self.regs[index] = value;
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/mod.rs:250]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/mod.rs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+        self.phase = self.phase.wrapping_add(1) & 0x1F;
+    }
+
+    fn step_noise_lfsr(&mut self) {
+        let divider = noise_clock_divider(self.regs[usize::from(FLG_REG)] & 0x1F);
+        self.noise_counter = self.noise_counter.wrapping_add(1);
+        if self.noise_counter < divider {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/mod.rs:192]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/envelope.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+fn envelope_tick_due(voice: &mut VoiceState) -> bool {
+    let rate = active_rate(voice);
+    let divider = rate_to_divider(rate);
+    voice.env_divider_counter = voice.env_divider_counter.wrapping_add(1);
+    if voice.env_divider_counter < divider {
+        return false;
+    }
+    voice.env_divider_counter = 0;
+    true
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/envelope.rs]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    for addr in 0u8..=0x7F {
+        let value = addr.wrapping_mul(3).wrapping_add(1);
+        dsp.write_reg(addr, value);
+    }
+
+    for addr in 0u8..=0x7F {
+        let value = addr.wrapping_mul(3).wrapping_add(1);
+        assert_eq!(dsp.read_reg(addr), value, "addr=0x{addr:02X}");
+    }
+}
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:34]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    assert_eq!(dsp.voice_sample_pos(2), 0x1234);
+}
+
+#[test]
+fn given_constant_samples_when_gaussian_interpolating_then_output_preserves_level() {
+    let dsp = Sdsp::new();
+    let out = dsp.gaussian_interpolate(100, 100, 100, 100, 64);
+    assert!((out - 100).abs() <= 2);
+    assert_eq!(out & 1, 0);
+}
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:83]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+fn given_non_voice_when_noise_clock_ticks_then_outx_changes_from_silence() {
+    let mut dsp = Sdsp::new();
+    dsp.write_reg(0x0C, 127);
+    dsp.write_reg(0x1C, 127);
+    dsp.write_reg(0x00, 127);
+    dsp.write_reg(0x01, 127);
+    dsp.write_reg(0x07, 0x7F); // direct gain for non-zero envelope
+    dsp.write_reg(0x3D, 0x01); // NON voice 0
+    dsp.write_reg(0x6C, 0x00); // fastest noise clock in this implementation
+
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:208]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+#[test]
+fn given_pmon_enabled_for_voice1_when_voice0_outx_nonzero_then_voice1_pitch_step_is_modulated() {
+    let mut dsp = Sdsp::new();
+    dsp.write_reg(0x0C, 127);
+    dsp.write_reg(0x1C, 127);
+    dsp.write_reg(0x00, 127);
+    dsp.write_reg(0x01, 127);
+    dsp.write_reg(0x08, 0x7F);
+    dsp.write_reg(0x12, 0x00); // voice1 pitch low
+    dsp.write_reg(0x13, 0x10); // voice1 pitch high => 0x1000
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:229]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    let base_step = dsp.voice_sample_pos(1);
+
+    let mut modulated = Sdsp::new();
+    modulated.write_reg(0x0C, 127);
+    modulated.write_reg(0x1C, 127);
+    modulated.write_reg(0x00, 127);
+    modulated.write_reg(0x01, 127);
+    modulated.write_reg(0x08, 0x7F);
+    modulated.write_reg(0x12, 0x00);
+    modulated.write_reg(0x13, 0x10);
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:241]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+#[test]
+fn given_pmon_enabled_when_master_volume_zero_then_pitch_modulation_still_applies() {
+    let mut base = Sdsp::new();
+    base.write_reg(0x00, 127);
+    base.write_reg(0x01, 127);
+    base.write_reg(0x08, 0x7F);
+    base.write_reg(0x12, 0x00);
+    base.write_reg(0x13, 0x10);
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:296]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.

File: src/snes/apu/dsp/tests.rs [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    base.step_phase();
+    let base_step = base.voice_sample_pos(1);
+
+    let mut modulated = Sdsp::new();
+    modulated.write_reg(0x00, 127);
+    modulated.write_reg(0x01, 127);
+    modulated.write_reg(0x08, 0x7F);
+    modulated.write_reg(0x12, 0x00);
+    modulated.write_reg(0x13, 0x10);
+    modulated.write_reg(0x0C, 0x00);
```
[the remark this answers is not in the document]
Comment (rmstdope) [reply src/snes/apu/dsp/tests.rs:306]: Addressed in commit a2c9c545.

Applied fixes for the reviewed concern in the S-DSP implementation/tests and re-ran focused DSP tests plus strict clippy. Thanks for the detailed review.
